Add support for DECIMAL/NUMERIC in JDBC connectors

findepi opened the pull request.

losipiuk commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -484,9 +485,10 @@ protected Type toPrestoType(int jdbcType, int columnSize, int decimalDigits)
             case Types.NUMERIC:
             case Types.DECIMAL:
-                return DOUBLE;
+                return createDecimalType(columnSize, decimalDigits);
```
What about the case when we exceed precision/scale supported by Presto?
Maybe we should stay with `DOUBLE` then?

findepi replied:
🤔  ... As we support finite precision and no negative scale (e.g. no support for `decimal(40, -2)`), this is the only option. 👍

losipiuk replied:
I see other options. As for throw an error otherwise. Or make it configurable. I am a little concerned about braking backward compatibility here. But maybe this is not a huge problem.

findepi commented:
> What about updating presto-mysql connector - no decimal there? Same for presto-sqlserver.

I added decimal support in `presto-base-jdbc`, so this work out of the box. However, I added a test in somewhat arbitrarily chosen connector (postrgesql), cause it's not possible to test base-jdbc directly.
Perhaps I could make a generic test (a la `ATQ` for jdbc connectors) and let postgresql,mysql,sqlserver all leverage it. What do you think?

losipiuk commented:
>I added decimal support in presto-base-jdbc, so this work out of the box. However, I added a test in somewhat arbitrarily chosen connector (postrgesql), cause it's not possible to test base-jdbc directly.
Perhaps I could make a generic test (a la ATQ for jdbc connectors) and let postgresql,mysql,sqlserver all leverage it. What do you think?

Seems reasonable. Maybe sth specific for type mapping testing. `base-jdbc/.../BaseTypeMappingTest.java`?
- with overridable methods like: `testMappingForDouble`, `testMappingForVarchar` ... and also `testReverseMappingFor...`.

I do not think it must go in this PR as there are not really tests for mapping for other types and other connectors so this is a lot of work.
File an issue?

findepi commented:
@losipiuk i created #9290 for the better tests.

findepi commented:
@losipiuk back to you.

* negative scale is now mapped to `decimal(precision + abs(scale), 0)`
* precision exceeding max supported is explicitly not handled (column not mapped)

Both things are covered by a read path test.

losipiuk commented:
please squash fixups

findepi commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -523,6 +535,9 @@ protected String toSqlType(Type type)
             }
             return "char(" + ((CharType) type).getLength() + ")";
         }
+        if (type instanceof DecimalType) {
```
i should also have a test when table created in Presto

losipiuk commented on presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java:
```diff
@@ -86,12 +90,148 @@ public void testForeignTable()
+        try {
+            // Check created column data type with PostgreSQL
+            String actualDataType = execute(statement -> {
+                try (ResultSet resultSet = statement.executeQuery(
```
Really? Cant you do that simpler?

Like here for example? https://stackoverflow.com/questions/2146705/select-datatype-of-the-field-in-postgres

findepi replied:
This is simplest way I know of that pulls type _and_ precision/scale/size information. Regarding the link, you can upvote best answer there (i.e. mine:)

I am, however, worried that for other database this won't be possible at all (no `format_type` function) and I will have either write awful `switch` in SQL or (rather) pull separate pieces of information from `information_schema`. However, these considerations I would like to leave for https://github.com/prestodb/presto/issues/9290
